lapack/testlapack/dgesvd.go

```diff
@@ -21,251 +23,351 @@ type Dgesvder interface {
+// according to mtype as:
+//  - the zero matrix if mtype == 1,
+//  - the identity matrix if mtype == 2,
+//  - a random matrix with a given condition number and singular values if mtype == 3, 4, or 5.
+// It first computes the full SVD  A = U*Sigma*V^T  and checks that
+//  - U has orthonormal columns, and V^T has orthonormal rows,
+//  - U*Sigma*V^T multiply back to A,
+//  - the singular values are non-negative and sorted in decreasing order.
+// Then all combinations of partial SVD results are computed and checked whether
+// they match the full SVD result.
+func dgesvdTest(t *testing.T, impl Dgesvder, m, n, mtype int) {
+	const tol = 1e-13
+
 	rnd := rand.New(rand.NewSource(1))
-	// TODO(btracey): Add tests for all of the cases when the SVD implementation
-	// is finished.
-	// TODO(btracey): Add tests for m > mnthr and n > mnthr when other SVD
-	// conditions are implemented. Right now mnthr is 5,000,000 which is too
-	// large to create a square matrix of that size.
-	for _, test := range []struct {
-		m, n, lda, ldu, ldvt int
-	}{
-		{5, 5, 0, 0, 0},
-		{5, 6, 0, 0, 0},
-		{6, 5, 0, 0, 0},
-		{5, 9, 0, 0, 0},
-		{9, 5, 0, 0, 0},
 
-		{5, 5, 10, 11, 12},
-		{5, 6, 10, 11, 12},
-		{6, 5, 10, 11, 12},
-		{5, 5, 10, 11, 12},
-		{5, 9, 10, 11, 12},
-		{9, 5, 10, 11, 12},
+	// Use a fixed leading dimension to reduce testing time.
+	lda := n + 3
+	ldu := m + 5
+	ldvt := n + 7
 
-		{300, 300, 0, 0, 0},
-		{300, 400, 0, 0, 0},
-		{400, 300, 0, 0, 0},
-		{300, 600, 0, 0, 0},
-		{600, 300, 0, 0, 0},
+	minmn := min(m, n)
 
-		{300, 300, 400, 450, 460},
-		{300, 400, 500, 550, 560},
-		{400, 300, 550, 550, 560},
-		{300, 600, 700, 750, 760},
-		{600, 300, 700, 750, 760},
-	} {
-		jobU := lapack.SVDAll
-		jobVT := lapack.SVDAll
+	// Allocate A and fill it with random values. The in-range elements will
+	// be overwritten below according to mtype.
+	a := make([]float64, m*lda)
+	for i := range a {
+		a[i] = rnd.NormFloat64()
+	}
 
-		m := test.m
-		n := test.n
-		lda := test.lda
-		if lda == 0 {
-			lda = n
+	var aNorm float64
+	switch mtype {
+	default:
+		panic("unknown test matrix type")
+	case 1:
+		// Zero matrix.
+		for i := 0; i < m; i++ {
+			for j := 0; j < n; j++ {
+				a[i*lda+j] = 0
+			}
 		}
-		ldu := test.ldu
-		if ldu == 0 {
-			ldu = m
+		aNorm = 0
+	case 2:
+		// Identity matrix.
+		for i := 0; i < m; i++ {
+			for j := 0; j < n; j++ {
+				if i == j {
+					a[i*lda+i] = 1
+				} else {
+					a[i*lda+j] = 0
+				}
+			}
 		}
-		ldvt := test.ldvt
-		if ldvt == 0 {
-			ldvt = n
+		aNorm = 1
+	case 3, 4, 5:
+		// Scaled random matrix.
+		// Generate singular values.
+		s := make([]float64, minmn)
+		// 4 means that s[i] = 1 - i*(1-1/cond)/(minmn-1),
+		// where cond = max(1, minmn),
+		// false means that signs of s[i] are not randomly flipped,
+		// 1 means that the random numbers are drawn uniformly from [0,1).
+		Dlatm1(s, 4, float64(max(1, minmn)), false, 1, rnd)
```

Comment:
You could do this as (only suggestion):
```
Dlatm1(s,
    4,                      // s[i] = 1 - i*(1-1/cond)/(minmn-1),
    float64(max(1, minmn)), // where cond = max(1, minmn)
    false,                  // signs of s[i] are not randomly flipped
    1, rnd)                 // random numbers are drawn uniformly from [0,1)
```

Reply:
Nice, thanks.

One day I will split `Dlatm1` into several smaller single-purpose functions or at least unexport it. The signature is really bad.